### PR TITLE
Bsg Wormhole Streamer Modules

### PR DESCRIPTION
**Pull request by dpetrisko**
These haven't been completely tested, but I'm making the PR for preliminary feedback.

The idea is to prevent wormhole serialization/deserialization penalties in a flexible, extendible way.  This is done by streaming in #header flits, allowing external logic to determine what to do, and then streaming in #data flits, allowing external logic to stream the data unserialized.

A few notes:
- This still requires a wormhole encoder module for specific protocol logic
- Flit width is decoupled from header width and data width; however, both header and data need to be a multiple of flit width at the moment.  For headers, this is a difficult problem to solve.  For data, we could introduce padding
- There may be bubbles introduced in this first implementation, although I believe they can be totally eliminated with some care

### CODE REVIEW

**Comment by taylor-bsg** [comment]
Looks interesting!

**Comment by gaozihou** [comment]
Hi Dan, I think in general the module interface looks great, it is straightforward and easy to use. For step 1 we assume data_width is multiple of wormhole_flit_width, later on we could extend the usage and allow arbitrary hdr and data width. (Current module interface allows us to do this)

**Line-to-line comments:**
I think streamer_out looks nice, just not completed yet
For streamer_in, I have two thoughts:
1. Currently state machine is controlling both before and after piso/sipo. If written correctly this will cause 2 cycles of bubble for each wormhole packet transmitted (in first cycle of "is_hdr" or "is_data" it does not have any wormhole flit sent out). Also it requires the parallel interface be "ready-then-valid", which (I think) is not desirable. I guess we could let state machine control wormhole-link side only (just like what we have in streamer_out), in this way it is bubble-free and parallel interface can naturally become ready_valid_and. (counter parameter set_and_down_exclusive_p=0 will be needed, because we extract wh_hdr after PISO/SIPO)
2. The state-machine should jump from is_hdr to is_data based on count_r_o (this is the main issue that make it not passing the test). Also, instead of counting from len+1 to 0, we could count from len to 0 to get rid of one-cycle bubble.

File: bsg_noc/bsg_wormhole_stream_in.v
Comment (dpetrisko) [inline bsg_noc/bsg_wormhole_stream_in.v]: Why is the condition data_len_li in one and data_flit_done in the other?

Comment (gaozihou) [reply bsg_noc/bsg_wormhole_stream_in.v]: When there are multiple hdr flits, data_len_li only meaningful in first hdr flit, so for multi-hdr-flit case we use the registered num_data_flit value (data_counter value) to see whether it is zero-data-flit case.

Comment (dpetrisko) [reply bsg_noc/bsg_wormhole_stream_in.v]: Got it!

Comment (gaozihou) [reply bsg_noc/bsg_wormhole_stream_in.v]: Thanks Dan! Comments added to clarify this piece of code.

File: bsg_noc/bsg_wormhole_stream_out.v
Comment (dpetrisko) [inline bsg_noc/bsg_wormhole_stream_out.v]: Don't these need to be link accept? The incoming link is rv& so even if data_ready_lo is low, data_v_li could be high

File: bsg_noc/bsg_wormhole_stream_out.v
Comment (dpetrisko) [inline bsg_noc/bsg_wormhole_stream_out.v]: nit, avoid

File: bsg_noc/bsg_wormhole_stream_out.v
Comment (dpetrisko) [inline bsg_noc/bsg_wormhole_stream_out.v]: Should this comment be link_v_i instead of hdr_v_lo?

File: bsg_noc/bsg_wormhole_stream_out.v
Comment (dpetrisko) [inline bsg_noc/bsg_wormhole_stream_out.v]: I guess it doesn't matter because we'll just set repeatedly until ready, but is there a reason to use link_v_i instead of link_accept here?

File: bsg_noc/bsg_wormhole_stream_out.v
Comment (gaozihou) [reply bsg_noc/bsg_wormhole_stream_out.v]: https://github.com/bespoke-silicon-group/basejump_stl/blob/c10593e4349a5ca1dfa1ac2b1c0d43d787ba3278/bsg_dataflow/bsg_parallel_in_serial_out.v#L35

This line of comment is out-dated :) It is actually rv&

File: bsg_noc/bsg_wormhole_stream_out.v
Comment (gaozihou) [reply bsg_noc/bsg_wormhole_stream_out.v]: Thanks Dan!

File: bsg_noc/bsg_wormhole_stream_out.v
Comment (gaozihou) [reply bsg_noc/bsg_wormhole_stream_out.v]: Yeah... Forgot to modify comment. 
Actually I am working on a module named "bsg_wormhole_stream_control", which includes the state-machine and counters, so that stream_in and stream_out can instantiate that module to prevent copy-n-paste large portion of code.

File: bsg_noc/bsg_wormhole_stream_out.v
Comment (gaozihou) [reply bsg_noc/bsg_wormhole_stream_out.v]: You are right, link_v_i and link_accept would both function correctly (link_v_i updates counter immediately, link_accept updates counter in last cycle). 

I chose link_v_i for no specific reason... Maybe we should use link_accept to make the code more readable?

Comment (dpetrisko) [reply bsg_noc/bsg_wormhole_stream_out.v]: Works for me!

File: bsg_noc/bsg_wormhole_stream_out.v
Comment (dpetrisko) [reply bsg_noc/bsg_wormhole_stream_out.v]: Cool!

File: bsg_noc/bsg_wormhole_stream_out.v
Comment (dpetrisko) [reply bsg_noc/bsg_wormhole_stream_out.v]: Oh, okay!

**Comment by dpetrisko** [comment]
Looks great to me! I'll try to write some better tests

**Comment by dpetrisko** [comment]
Added a few tests.  Achieving 90% line and 80% toggle coverage, probably can get more by sweeping parameters (the tests only use one set of parameters).

**Comment by dpetrisko** [comment]
Hey @gaozihou, did you just change parameters in the trace gen manually? We should add the different tests to the testing/ Makefile and then I think we’re good to go on this

**Comment by dpetrisko** [comment]
Any objections to merging this?

File: bsg_noc/bsg_wormhole_stream_control.v
Comment (taylor-bsg) [inline bsg_noc/bsg_wormhole_stream_control.v]: High level descriptions of semantics and design overview for each of the three modules in comments is in order  =) for such a large bunch of code; or if there is a google doc, link to it.

Comment (dpetrisko) [reply bsg_noc/bsg_wormhole_stream_control.v]: Haha guilty. I'll write up the many whiteboard drawings Paul and I went through :)

**Comment by dpetrisko** [comment]
Updated with documentation and high level descriptions of the modules.  Good to go?

**Comment by dpetrisko** [comment]
Bump

**Comment by dpetrisko** [comment]
@taylor-bsg Is there anything else that needs to be done for these modules?

**Comment by dpetrisko** [comment]
@taylor-bsg Can we merge this?  Needed for BlackParrot: https://github.com/black-parrot/black-parrot/pull/618

**Review by tommydcjung — APPROVED** [review]
I don't see any major issues here. lgtm

File: bsg_noc/bsg_wormhole_stream_control.v
Comment (dpetrisko) [reply bsg_noc/bsg_wormhole_stream_control.v]: fixed!

File: bsg_noc/bsg_wormhole_stream_in.v
Comment (taylor-bsg) [inline bsg_noc/bsg_wormhole_stream_in.v:26]: Clarify what is illegal as a comment e.g., (Header sent too early)

File: bsg_noc/bsg_wormhole_stream_in.v
Comment (taylor-bsg) [inline bsg_noc/bsg_wormhole_stream_in.v:36]: Define cid

File: bsg_noc/bsg_wormhole_stream_in.v
Comment (taylor-bsg) [inline bsg_noc/bsg_wormhole_stream_in.v:52]: hdr_ready_and_o?

File: bsg_noc/bsg_wormhole_stream_in.v
Comment (taylor-bsg) [inline bsg_noc/bsg_wormhole_stream_in.v:57]: data_ready_and_o?

File: bsg_noc/bsg_wormhole_stream_in.v
Comment (taylor-bsg) [inline bsg_noc/bsg_wormhole_stream_in.v:62]: link_ready_and_i?

File: bsg_noc/bsg_wormhole_stream_in.v
Comment (taylor-bsg) [inline bsg_noc/bsg_wormhole_stream_in.v:74]: seems like maybe it is worth special casing if hdr_len_lp == 1? If I understand correctly, both the output and input are ready/valid, so the piso could be replaced with pass through wires? if so, I might make is_hdr be is_hdr_r, coming from is_hdr_r_o from the bsg_wormhole_stream_control?

Comment (dpetrisko) [reply bsg_noc/bsg_wormhole_stream_in.v:74]: Special case of hdr_len_lp == 1 degenerates into a two fifo.  Just passing through could result in an unexpected timing path to a router?

File: bsg_noc/bsg_wormhole_stream_in.v
Comment (taylor-bsg) [inline bsg_noc/bsg_wormhole_stream_in.v:126]: comment as to why this setting is appropriate

File: bsg_noc/bsg_wormhole_stream_in.v
Comment (taylor-bsg) [inline bsg_noc/bsg_wormhole_stream_in.v:7]: I think more exposition on each expected use cases would be useful. For example, there are a bunch of limitations on the relation of data width and flit width, so I think that probably implies that there are certain relationships between the input and output wormhole networks that must be obeyed?  When you go between two wormhole networks, do you need to manually translate the len field of the header between the two networks? etc.

File: bsg_noc/bsg_wormhole_stream_control.v
Comment (taylor-bsg) [reply bsg_noc/bsg_wormhole_stream_control.v:1]: Is there a link to a google doc somewhere here with said whiteboard drawings?

File: bsg_noc/bsg_wormhole_stream_in.v
Comment (taylor-bsg) [reply bsg_noc/bsg_wormhole_stream_in.v:74]: after you update the interface with ready_and versus ready_then, then I can answer that question better.. =)

Comment (taylor-bsg) [reply bsg_noc/bsg_wormhole_stream_in.v:79]: Which paths are you thinking of? Presumably the ready/valid signals are okay, since if they weren't, the interface is incorrect for the module. For the data signal, it doesn't seem like the logic depth is particularly deep; but let me ask this question; what is the intended use of this module? Is the input connected to an input FIFO coming from a router? Is the output going to the local input of a router, or is it intended to go a long distance across the chip?

File: bsg_noc/bsg_wormhole_stream_in.v
Comment (taylor-bsg) [reply bsg_noc/bsg_wormhole_stream_in.v:133]: but in this system, you can have multiple output data flits associated with a header, right? So you will have a bubble between data flits?

File: bsg_noc/bsg_wormhole_stream_in.v
Comment (dpetrisko) [reply bsg_noc/bsg_wormhole_stream_in.v:7]: Added use cases and some description of the width limitation.  The width limitations are entirely driven by the SIPO/PISO-based implementation.  Because this module completely decouples the networks on both sides, a v2 could easily support non-flit-multiple header/data widths

File: bsg_noc/bsg_wormhole_stream_in.v
Comment (dpetrisko) [reply bsg_noc/bsg_wormhole_stream_in.v:90]: this module will most likely be on the input P-port of a router, locally connected to some driver (bsg_cache dma, etc.).  So for stream_in it will be connected to the input fifo of the router. I think you're right that that's probably fine.

for bsg_wormhole_stream_out, it will be on the output P-port of a router, so no input FIFO.  However, this also probably fine as long as the client is latching inputs as is standard in BaseJump

File: bsg_noc/bsg_wormhole_stream_in.v
Comment (dpetrisko) [reply bsg_noc/bsg_wormhole_stream_in.v:144]: The bubble happens between consecutive bursts, not consecutive words.

  header  -  data   -  header  -   data
(bubble) A A A A A (bubble) B B B B B

File: bsg_noc/bsg_wormhole_stream_control.v
Comment (dpetrisko) [reply bsg_noc/bsg_wormhole_stream_control.v:1]: No google doc, and I think the whiteboard drawings are lost to the ravages of time

File: bsg_noc/bsg_wormhole_stream_in.v
Comment (dpetrisko) [reply bsg_noc/bsg_wormhole_stream_in.v]: updated

Comment (gaozihou) [reply bsg_noc/bsg_wormhole_stream_in.v]: I think passthrough may cause one cycle bubble between hdr and data flit. One example is bsg_cache_dma.v: https://github.com/bespoke-silicon-group/basejump_stl/blob/bsg_wormhole_streamer/bsg_cache/bsg_cache_dma.v, the "single hdr flit" (dma_pkt) is passthrough, and data flits has two_fifo on output. I had to add an extra two_fifo on dma_pkt port outside bsg_cache_dma module to eliminate the bubble...

The bubble appears when hdr flit and data flit cannot arrive at stream_in module in same cycle (which is not rare, like when we use a state-machine to generate the traffic, we send out hdr in state0, and send out data in state1, then go back to state0...). Preserving the extra buffer for hdr_flit will make this module easier to use (not likely to introduce bubble)

File: bsg_noc/bsg_wormhole_stream_in.v
Comment (gaozihou) [reply bsg_noc/bsg_wormhole_stream_in.v:161]: I think "use_minimal_buffering_p==1" may introduce bubble, because there might be multiple data flits coming out continuously. Output data of SIPOF are not truly-parallel, they are "wider" data flits that run into wormhole link.

For example:
`header - data - (bbl) - data - header - data - (bbl) - data`
`(bbl) - AAAAA - (bbl) - AAAAA - (bbl) - BBBBB - (bbl) - BBBBB`
In this case there will be extra bubble between two data flits.

Comment (dpetrisko) [reply bsg_noc/bsg_wormhole_stream_in.v:161]: Oh you're right

Comment (dpetrisko) [reply bsg_noc/bsg_wormhole_stream_in.v]: removed minimal buffering

File: bsg_noc/bsg_wormhole_stream_in.v
Comment (dpetrisko) [reply bsg_noc/bsg_wormhole_stream_in.v]: Not sure I understand.  Yes, there's a bubble between the header and data, but that's no less latency than having the header in a fifo and come out a cycle later. Well, I guess it would tie up the router port for an extra cycle.  But it would free up the endpoint to continue computing.

I think it's safe with all combinations of passthrough and not passthrough.  Because we have the stream control module arbitrate based on is_hdr and is_data, data coming a cycle early should not cause issues.  I could be wrong though.

That's an interesting point I hadn't thought of though, daisy chaining these modules could unacceptably increase critical path on passthrough.

Could you explain the dma_pkt problem a little more in depth?

Comment (gaozihou) [reply bsg_noc/bsg_wormhole_stream_in.v]: Absolutely.

Diagram above shows the cache_dma design. Imagine that the outside world is trying to merge dma_hdr and dma_data traffic (like going into a SIPO, inside cache_dma_to_cce adapter)

Cycle 0 - send out dma_hdr - outside world receive dma_hdr
Cycle 1 - send out data0 - outside world receive nothing (bubble cycle)
Cycle 2 - send out data1 - outside world receive data0
...
Cycle 8 - send out data7 - outside world receive data6
Cycle 9 - cannot send - outside world receive data7
Cycle 10 - send out dma_hdr - outside world receive dma_hdr
...

If we add a two_fifo on dma_hdr, the bubble can be eliminated:

Cycle 0 - send out dma_hdr - outside world receive nothing
Cycle 1 - send out data0 - outside world receive dma_hdr
Cycle 2 - send out data1 - outside world receive data0
...
Cycle 8 - send out data7 - outside world receive data6
Cycle 9 - send out dma_hdr - outside world receive data7
Cycle 10 - send out data0- outside world receive dma_hdr
...

Comment (dpetrisko) [reply bsg_noc/bsg_wormhole_stream_in.v]: I guess my point is it's still 10 cycles right? As an unrelated aside, seems like we should add a one/two fifo to the dma hdr?

Comment (gaozihou) [reply bsg_noc/bsg_wormhole_stream_in.v]: After adding two_fifo it is 9 cycles (see the updated cycle counts)

Comment (dpetrisko) [reply bsg_noc/bsg_wormhole_stream_in.v]: Ah, I see!

Comment (dpetrisko) [reply bsg_noc/bsg_wormhole_stream_in.v]: I'm fine with changing it back, @taylor-bsg ?

**Comment by taylor-bsg** [comment]
I have some extended comments that are in a desktop browser window, but I
think the high level takeaway is that our implementation has issues because
the interface to piso assumes ready_and on input and yumi on output, which
requires built in FIFO buffering; but this module is double ready_and and
doesn’t need it. The result is a bunch of unnecessary FIFO plumbing and
related complexity that should not be in the module. A pass thru PISO would
simplify this problem.

**Comment by dpetrisko** [comment]
>The result is a bunch of unnecessary FIFO plumbing and
related complexity that should not be in the module. A pass thru PISO would
simplify this problem.

Do you mean a PISO with no internal buffering at all, so an interface like:
```
clk_i
,reset_i

,data_i
,v_i
,yumi_o

,data_o
,v_o
,yumi_i
```

where a N:1 PISO would send out 1 yumi_o for every N yumi_i's?

**Comment by gaozihou** [comment]
> Do you mean a PISO with no internal buffering at all, so an interface like:
> 
> ```
> clk_i
> ,reset_i
> 
> ,data_i
> ,v_i
> ,yumi_o
> 
> ,data_o
> ,v_o
> ,yumi_i
> ```
> 
> where a N:1 PISO would send out 1 yumi_o for every N yumi_i's?

I think it is a PISO that has ready_and_i on one side and ready_and_o on the other side, without any internal buffering. So do SIPOF.

**Comment by taylor-bsg** [comment]
Dan: yes, except it is ready_and on both sides
Paul: yes, except I don’t know what SIPOF is

**Comment by taylor-bsg** [comment]
FYI I think bsg_round_robin_n_to_1 is essentially Dan’s variant.

**Comment by dpetrisko** [comment]
>Dan: yes, except it is ready_and on both sides

Not sure I understand the semantics here.  If we have no buffering, what does (ready_and_o & v_i) on piso input mean? I can see two possibilities:

1) ready_and_o means that we're ready to accept a new request

Then we need an internal bsg_dff_en_bypass to store the wide data, because once the (ready_and_o & valid_i) are high, the data is acked and gone.

2) ready_and_o means that we're done transmitting.  Then it's more natural as a yumi_o, because the valid goes high when there's new data, and yumi goes high when the data is transmitted.

Which do you mean?

**Comment by taylor-bsg** [comment]
On Wed, Sep 30, 2020 at 9:59 PM Dan Petrisko <notifications@github.com>
wrote:

> Dan: yes, except it is ready_and on both sides
>
> Not sure I understand the semantics here. If we have no buffering, what
> does (ready_and_o & v_i) on piso input mean? I can see two possibilities:
>
>    1. ready_and_o means that we're ready to accept a new request
>
> Then we need an internal bsg_dff_en_bypass to store the wide data, because
> once the (ready_and_o & valid_i) are high, the data is acked and gone.
>
>    1. ready_and_o means that we're done transmitting. Then it's more
>    natural as a yumi_o, because the valid goes high when there's new data, and
>    yumi goes high when the data is transmitted.
>
> Which do you mean?
>

No internal data storage.

From the perspective of the piso:

- v_i means there is upstream data available
- ready_and_o, as in all cases, means that if there was a request (v_i), we
accept it.
- ready_and_i means the downstream module is willing to accept a single
word.
- v_o means that a single word is available for the downstream module to
absorb

practically speaking:
    v_o = v_i
   ready_and_o = ready_and_i & last_word_r
last_word_r is updated based on v_o & ready_and_i

A yumi_o signal is dependent on the incoming valid_i signal of that
interface; while a ready_and_o is not...

**Comment by dpetrisko** [comment]
Updated with passthrough sipo/pisos.  Depends on #345 #346

File: bsg_dataflow/bsg_serial_in_parallel_out_passthrough.v
Comment (taylor-bsg) [inline bsg_dataflow/bsg_serial_in_parallel_out_passthrough.v:54]: I think only the last element needs to be bypassed?

Comment (dpetrisko) [reply bsg_dataflow/bsg_serial_in_parallel_out_passthrough.v:54]: Oh great point!

Comment (gaozihou) [reply bsg_dataflow/bsg_serial_in_parallel_out_passthrough.v:54]: The last element may not need to be registered (if we prefer minimum buffering)

Comment (dpetrisko) [reply bsg_dataflow/bsg_serial_in_parallel_out_passthrough.v:54]: even better :)

Comment (dpetrisko) [reply bsg_dataflow/bsg_serial_in_parallel_out_passthrough.v:54]: Fixed in the SIPO PR

File: bsg_dataflow/bsg_serial_in_parallel_out_passthrough.v
Comment (taylor-bsg) [inline bsg_dataflow/bsg_serial_in_parallel_out_passthrough.v:55]: misplaced comment

**Comment by dpetrisko** [comment]
Updated again with updated passthrough sipo/pisos. Depends on #345 #346

**Comment by dpetrisko** [comment]
Okay, I think this is finally good to go.  I updated the handshakes to be ready_and throughout to match the passthrough piso/sipos.  All tests pass for wormhole_stream, wormhole_stream_in and wormhole_stream_out.

File: bsg_noc/bsg_wormhole_stream_in.v
Comment (muwyse) [inline bsg_noc/bsg_wormhole_stream_in.v:149]: port is currently named ready_o not ready_and_o

File: bsg_noc/bsg_wormhole_stream_in.v
Comment (muwyse) [inline bsg_noc/bsg_wormhole_stream_in.v:153]: similarly, port currently yumi_i not ready_and_i

Comment (dpetrisko) [reply bsg_noc/bsg_wormhole_stream_in.v:153]: fixed

**Comment by dpetrisko** [comment]
@taylor-bsg can we get a final approval on this?

**Comment by muwyse** [comment]
@taylor-bsg any timeline on this merge? This is currently blocking BP PR 626.

**Comment by dpetrisko** [comment]
Regarding the wormhole_router arbitration issue we discussed on Tuesday, I think we can merge this before fixing it.  We have an assertion that the data/v_i must not change during the PISO transaction, so if this condition does arise, we'll immediately know what the problem is.

Should make a separate issue to add the locking fixed arbiter as a parameter to wormhole_router though

**Comment by dpetrisko** [comment]
@taylor-bsg 

It seems there actually isn't a problem with the wormhole streamers + wormhole routers as currently implemented.
The arbitration issue only happens when the RR switches in the middle of header transmission, not data

- For router -> streamer connection, the header can be 1 or multiple flits.  The header is aggregated by a passthrough SIPO and doesn't have a problem

- For streamer -> router connection, the header can be multiple flits.  This is a passthrough PISO, so the header needs to be held constant.  However, the input to the PISO is connected to the core logic, not the router, so there is no switching

There would be a problem if we connected a round-robin arbiter on the core logic side of the streamer, but (at least in BP), we don't have a use-case for this, since we handle core-side arbitration with concentrators.  The assertion should be sufficient

**Comment by dpetrisko** [comment]
@taylor-bsg I opened #232. The problematic modules (bsg_mesh_router, bsg_parallel_in_serial_out_passthrough) are not included in this PR, the identified problem is not introduced in this PR, we have an assertion for unsupported usage, and there is not a simple fix.  We should merge this to unblock several major PPA improvements in BlackParrot

**Comment by dpetrisko** [comment]
@taylor-bsg @gaozihou I added an additional round-robin arbiter mode and a test.  Should be good to go

**Comment by dpetrisko** [comment]
Depends on #359

**Comment by dpetrisko** [comment]
Bugfix in #358

**Comment by dpetrisko** [comment]
Coming up on this PR's anniversary.  see round robin arbiter feature here: #359 , otherwise this should be good to go

**Pull request closed without merging**
